Use Cobra CLI instead of Go flag package

chrishajas opened the pull request:
This set of commits migrates the gpbackup and gprestore utilities to use Cobra (https://github.com/spf13/cobra), which supports POSIX-compliant flags and allows us to support more complex CLI features in the future. We continue to support single dashes for flags (-flag) for backwards compatibility. This set of commits only changes the help UI, there is no change in functionality of the utilities.

I did not modify gpbackup_helper in this PR, as it is not user facing. However, we can also migrate it to use Cobra if we want.

New help UI:
```
gpbackup is the parallel backup utility for Greenplum

Usage:
  gpbackup [flags]

Flags:
      --backup-dir string           The absolute path of the directory to which all backup files will be written
      --compression-level int       Level of compression to use during data backup. Valid values are between 1 and 9.
      --data-only                   Only back up data, do not back up metadata
      --dbname string               The database to be backed up
      --debug                       Print verbose and debug log messages
      --exclude-schema strings      Back up all metadata except objects in the specified schema(s). --exclude-schema can be specified multiple times.
      --exclude-table strings       Back up all metadata except the specified table(s). --exclude-table can be specified multiple times.
      --exclude-table-file string   A file containing a list of fully-qualified tables to be excluded from the backup
      --help                        Help for gpbackup
```

professor commented:
FTW. Great job team.

khuddlefish commented on backup/validate.go:
```diff
@@ -71,18 +72,17 @@ WHERE quote_ident(n.nspname) || '.' || quote_ident(c.relname) IN (%s)`, quotedTa
-	utils.CheckExclusiveFlags("metadata-only", "leaf-partition-data")
-	utils.CheckExclusiveFlags("metadata-only", "single-data-file")
-	utils.CheckExclusiveFlags("no-compression", "compression-level")
+func ValidateFlagCombinations(cmd *cobra.Command) {
```
Do we need to pass in the cobra.Command here or should we just pass in cmd.Flags() directly?

khuddlefish commented on backup/backup_suite_test.go:
```diff
@@ -6,50 +6,41 @@ package backup_test
-/* This function is a helper function to execute gpbackup and return a session
- * to allow checking its output.
- */
-func gpbackup() *gexec.Session {
```
Was this dead code?

khuddlefish commented on gpbackup.go:
```diff
@@ -3,13 +3,27 @@
+	var rootCmd = &cobra.Command{
+		Use:   "gpbackup",
+		Short: "gpbackup is the parallel backup utility for Greenplum",
+		Args:  cobra.NoArgs,
```
I think the Args keyword is optional. Is "NoArgs" actually doing something that we need?

chrishajas replied:
With the NoArgs string, it will error out if the user attempts to use a subcommand, whereas without the string it will simply ignore subcommands.

With the NoArgs string: 
```
➜  gpbackup git:(cobra_cli) ✗ gpbackup afds --dbname test
Error: unknown command "afds" for "gpbackup"
Usage:
  gpbackup [flags]
```

Without it: 
```
➜  gpbackup git:(cobra_cli) ✗ gpbackup afds --dbname test
20180523:17:29:53 gpbackup:chajas:chris_mac:051855-[INFO]:-Starting backup of database test
20180523:17:29:53 gpbackup:chajas:chris_mac:051855-[INFO]:-Backup Timestamp = 20180523172953
```

khuddlefish replied:
ok cool. let's keep it then

khuddlefish commented on gpbackup.go:
```diff
@@ -3,13 +3,27 @@
+		Run: func(cmd *cobra.Command, args []string) {
+			defer DoTeardown()
+			DoFlagValidation(cmd)
+			DoSetup()
```
DoInit?

khuddlefish commented on gprestore.go:
```diff
@@ -3,13 +3,27 @@
+		Short: "gprestore is the parallel restore utility for Greenplum",
+		Args:  cobra.NoArgs,
+		Run: func(cmd *cobra.Command, args []string) {
+			defer DoTeardown()
```
Should we defer DoTeardown() outside of this function? Chances are small, but maybe we could get an error during initialization or something.

chrishajas replied:
Hmm... Init is quite small currently, we will panic if anything goes wrong. If I run `gpbackup --help`, it ends up printing 
```
      --version                     Print version number and exit
      --with-stats                  Back up query plan statistics
20180523:18:02:28 gpbackup:chajas:chris_mac:070175-[INFO]:-Backup completed successfully
```
at the end since it goes into cleanup after.

khuddlefish commented on restore/validate.go:
```diff
@@ -160,14 +161,14 @@ func validateBackupFlagPluginCombinations() {
-	utils.CheckExclusiveFlags("exclude-schema", "exclude-table", "include-table", "exclude-table-file", "include-table-file")
-	utils.CheckExclusiveFlags("exclude-table", "exclude-table-file", "leaf-partition-data")
-	utils.CheckExclusiveFlags("metadata-only", "data-only")
+func ValidateFlagCombinations(cmd *cobra.Command) {
```
Same as above. Maybe pass in the flags directly?

chrishajas replied:
Yep, will fix.

khuddlefish commented on utils/flag.go:
```diff
@@ -69,3 +43,17 @@ func IsValidTimestamp(timestamp string) bool {
 }
+
+/*
+ * Convert arguements that contain a single dash to double dashes for backward
```
arguements -> arguments

khuddlefish commented:
Somewhat tangential, but can we now detect the difference between a default value and the user passing in a value that happens to be equal to the default? (like for compression level)

chrishajas commented:
@khuddlefish Yes, using https://godoc.org/github.com/spf13/pflag#FlagSet.Changed. I modified CheckExclusiveFlags to use that.

chrishajas commented:
@khuddlefish 

If an error occurs in DoInit:

Current behavior or behavior with defer DoTeardown at top:
```
➜  ~ gpbackup -dbname test
/Users/chajas/gpAdminLogs is a file, not a directory
panic: /Users/chajas/gpAdminLogs is a file, not a directory [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x10eed08]

goroutine 1 [running]:
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.GetLogPrefix(0x134f073, 0x7, 0x1512808, 0x2)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:155 +0x28
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.Warn(0x135b6f6, 0x24, 0xc42006b7f8, 0x1, 0x1)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:198 +0x79
github.com/greenplum-db/gpbackup/backup.DoCleanup.func1()
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/backup/backup.go:375 +0xc9
panic(0x12f6a60, 0x14eac50)
	/usr/local/Cellar/go/1.10/libexec/src/runtime/panic.go:505 +0x229
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.GetLogPrefix(0x134e6be, 0x5, 0x1512808, 0x10525c0)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:155 +0x28
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.Verbose(0x1352650, 0x11, 0x0, 0x0, 0x0)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:206 +0x79
github.com/greenplum-db/gpbackup/backup.DoCleanup()
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/backup/backup.go:380 +0x79
github.com/greenplum-db/gpbackup/backup.DoTeardown()
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/backup/backup.go:363 +0x9b
panic(0x12e05e0, 0xc4200a8980)
	/usr/local/Cellar/go/1.10/libexec/src/runtime/panic.go:505 +0x229
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.abort(0xc42006bdc8, 0x1, 0x1)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:290 +0x121
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.createLogDirectory(0xc4200dc240, 0x19)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:313 +0x41e
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.InitializeLogging(0x134f79a, 0x8, 0xc4200dc240, 0x19)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:105 +0x8a
```
Behavior after this PR:
```
➜  ~ gpbackup -dbname test
panic: /Users/chajas/gpAdminLogs is a file, not a directory

goroutine 1 [running]:
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.abort(0xc42006bda0, 0x1, 0x1)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:290 +0x121
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.createLogDirectory(0xc4200ca2c0, 0x19)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:313 +0x41e
github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog.InitializeLogging(0x1428d3d, 0x8, 0xc4200ca2c0, 0x19)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:105 +0x8a
github.com/greenplum-db/gpbackup/backup.DoInit(0xc4200d1180)
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/backup/backup.go:49 +0x7e
main.main()
	/Users/chajas/go/src/github.com/greenplum-db/gpbackup/gpbackup.go:25 +0x137
```

chrishajas commented:
Added commits addressing the feedback and also making `--version` work again. Let me know if there's anything else you think we should change before I merge.
